redis/options.go

```diff
@@ -0,0 +1,376 @@
+			}
+		}
+	case *sentinelOptions:
+	case *singleNodeOptions:
+		universalOpts = &redis.UniversalOptions{
+			DB:              o.Database,
+			Username:        o.Username,
+			Password:        o.Password,
+			MaxRetries:      o.MaxRetries,
+			MinRetryBackoff: time.Duration(o.MinRetryBackoff) * time.Millisecond,
+			MaxRetryBackoff: time.Duration(o.MaxRetryBackoff) * time.Millisecond,
+		}
+		if err := setSocketOptions(universalOpts, o.Socket); err != nil {
+			return nil, err
+		}
+	case *redis.Options:
+		universalOpts = &redis.UniversalOptions{
+			Addrs:              []string{o.Addr},
+			DB:                 o.DB,
+			Username:           o.Username,
+			Password:           o.Password,
+			MaxRetries:         o.MaxRetries,
+			MinRetryBackoff:    o.MinRetryBackoff,
+			MaxRetryBackoff:    o.MaxRetryBackoff,
+			DialTimeout:        o.DialTimeout,
+			ReadTimeout:        o.ReadTimeout,
+			WriteTimeout:       o.WriteTimeout,
+			PoolSize:           o.PoolSize,
+			MinIdleConns:       o.MinIdleConns,
+			MaxConnAge:         o.MaxConnAge,
+			PoolTimeout:        o.PoolTimeout,
+			IdleTimeout:        o.IdleTimeout,
+			IdleCheckFrequency: o.IdleCheckFrequency,
+		}
+	default:
+		panic(fmt.Sprintf("unexpected options type %T", options))
+	}
+
+	return universalOpts, nil
+}
+
+func setSocketOptions(opts *redis.UniversalOptions, sopts *socketOptions) error {
+	opts.Addrs = []string{fmt.Sprintf("%s:%d", sopts.Host, sopts.Port)}
+	opts.DialTimeout = time.Duration(sopts.DialTimeout) * time.Millisecond
+	opts.ReadTimeout = time.Duration(sopts.ReadTimeout) * time.Millisecond
+	opts.WriteTimeout = time.Duration(sopts.WriteTimeout) * time.Millisecond
+	opts.PoolSize = sopts.PoolSize
+	opts.MinIdleConns = sopts.MinIdleConns
+	opts.MaxConnAge = time.Duration(sopts.MaxConnAge) * time.Millisecond
+	opts.PoolTimeout = time.Duration(sopts.PoolTimeout) * time.Millisecond
+	opts.IdleTimeout = time.Duration(sopts.IdleTimeout) * time.Millisecond
+	opts.IdleCheckFrequency = time.Duration(sopts.IdleCheckFrequency) * time.Millisecond
+
+	if sopts.TLS != nil {
+		tlsCfg := &tls.Config{}
+		if len(sopts.TLS.CA) > 0 {
+			caCertPool := x509.NewCertPool()
+			for _, cert := range sopts.TLS.CA {
+				caCertPool.AppendCertsFromPEM([]byte(cert))
+			}
+			tlsCfg.RootCAs = caCertPool
+		}
+
+		if sopts.TLS.Cert != "" && sopts.TLS.Key != "" {
+			clientCertPair, err := tls.X509KeyPair([]byte(sopts.TLS.Cert), []byte(sopts.TLS.Key))
+			if err != nil {
+				return err
+			}
+			tlsCfg.Certificates = []tls.Certificate{clientCertPair}
+		}
+
+		opts.TLSConfig = tlsCfg
+	}
+
+	return nil
+}
+
+// Set UniversalOption values from single-node options, ensuring that any
+// previously set values are consistent with the new values. This validates that
+// multiple node options set when using cluster mode are consistent with each other.
+// TODO: Use generics and/or reflection to DRY?
+func setConsistentOptions(uopts *redis.UniversalOptions, opts *singleNodeOptions) error {
```

Comment:
This function and the one below it are... awful :nauseated_face:, but, unfortunately, unavoidable if we want to have a Node Redis-like JS API, since we need to translate it into the go-redis one.

I considered abandoning `redis.UniversalOptions`, and using all 3 option variants instead (`redis.Options`, `redis.FailoverOptions` and `redis.ClusterOptions`), but we would also need to have a translation for them, so it wouldn't help much.

What I think would help DRY this a bit is to use generics and/or reflection, but we would have to bump the current `go 1.17` in `go.mod` to >=1.18, which we can safely do, since k6 uses 1.19.

I did it this way just to get it done quickly, but I definitely don't want to have to maintain this mess...

Reply:
I've updated the go version, k6 dependency, and go-redis dependency on master, I'll look into using generics/reflection to improve the situation indeed 👍🏻

Reply:
It turns out that generics/reflection ended up not being needed here, since I was able to reuse the same `setConsistentOptions()` function for both cluster and single-node options. This function is still too large and repetitive simply because of the amount of options we need to translate, but this is not something that generics or reflection can help with. We'd need a more dynamic language for that :sweat_smile:

So I left this as is, and removed the previous `setConsistentOptionsRedis()` function. I added a TODO to break it apart in the future, so that we can remove the current `nolint`s.